make/compiler_flags

```diff
@@ -123,8 +123,12 @@ CPPFLAGS_SUNDIALS ?= -DNO_FPRINTF_OUTPUT $(CPPFLAGS_OPTIM_SUNDIALS) $(CXXFLAGS_F
 #CPPFLAGS_GTEST ?=
 
 
-## setup compiler flags
-CXXFLAGS_LANG ?= -std=c++1y
+ifeq ($(STAN_USE_CPP14),)
+  CXXFLAGS_LANG ?= -std=c++17
+else
+  $(warning "Because STAN_USE_CPP14 is set C++14 standard is requested. This will not be possible in the next release!")
+  CXXFLAGS_LANG ?= -std=c++1y
+endif
```

Comment:
I don't like that this is going to cause compiler errors for users who do not have c++17. Can we do something like the below? Idt this is very portable

```make
CPP_STANDARD := $(shell $(CXX) -x c++ -E -dM - < /dev/null | grep -oP '__cplusplus \K\d+')

# Convert the CPP_STANDARD into a format usable by Make's conditionals
IS_CPP_STANDARD_GREATER_THAN_2017 := $(shell [ $(CPP_STANDARD) -gt 201700 ] && echo 1 || echo 0)

# Conditional directive based on the CPP_STANDARD
ifeq ($(IS_CPP_STANDARD_GREATER_THAN_2017),1)
CXXFLAGS_LANG ?= -std=c++17
else
$(warning "Your compiler does not support C++17 which will be required in the next release! Please upgrade your compiler.")
CXXFLAGS_LANG ?= -std=c++1y
endif
```

Reply:
That won't actually detect if the compiler _supports_ c++17, that will detect if it is the _default_ or not.

I agree it would be better if we can detect it, but I haven't found any reasonable way of doing so besides doing something like grepping `g++ -v --help` (and I have no idea if that would work with clang)

Reply:
Hmmm, what if we used CXX_MAJOR and CXX_MINOR to figure that out? We know what version of gcc and clang will support c++17. If someone is using neither of those we can just still have the way to downgrade to C++14 if they have an old compiler.

Then the only users whose code fails will be people not using gcc/clang

Reply:
I think some combination of `TBB_CXX_TYPE`, `CXX_MAJOR`, and `CXX_MINOR` could be used, yes. We'd need to know the exact values we want for each compiler

Reply:
I think it might be a bit more robust/portable to test for [a feature macro](https://en.cppreference.com/w/cpp/feature_test) for the corresponding support level

For example:
```sh
> echo | g++ -std=c++14 -x c++ -E -dM - | grep __cpp_aggregate_bases
> echo | g++ -std=c++17 -x c++ -E -dM - | grep __cpp_aggregate_bases
#define __cpp_aggregate_bases 201603L
```

Reply:
I think the second command there will fail if `-std=c++17` isn't a recognized standard by the compiler

Reply:
Still usable in a makefile though. 

Given dummy makefile:

```make
CXX17_CHECK := $(shell echo | $(CXX) -std=c++17 -x c++ -E -dM - 2>&1 | grep __cpp_aggregate_bases)
CXX30_CHECK := $(shell echo | $(CXX) -std=c++30 -x c++ -E -dM - 2>&1 | grep __cpp_aggregate_bases)

std_test:
	@echo "$(CXX17_CHECK)"
	@echo "$(CXX30_CHECK)"
	@echo "end"
```

Result is:

```sh
> make std_test
#define __cpp_aggregate_bases 201603L

end
```

Reply:
@andrjohns that seems better - could you take on writing the check?

My other concern is similar to when we added something like that for UCRT detection, which is if it will noticeably slow down compilation. If so we'd want to make a `make/cpp17` file or something that stores the result?

Reply:
@WardBrian could you check out the most recent change I made to this PR? I added some checks for each compiler version and if it is a version we know supports c++17 we use that as the standard

Reply:
I think the basic idea would work, but I'm pretty sure the logic as you have it written leads to CXXFLAGS_LANG just being completely unset if the user has a compiler type we recognize but it's too old?

Reply:
I'm also not confident that the `[ ]` program works on Windows tbh

Reply:
> I think the basic idea would work, but I'm pretty sure the logic as you have it written leads to CXXFLAGS_LANG just being completely unset if the user has a compiler type we recognize but it's too old?

Oh, yes!

It looks like the tests are running on windows for jenkins?

https://github.com/stan-dev/math/actions/runs/9116147777/job/25064095453?pr=3063#step:9:23

Reply:
I think the logic now is correct but overcomplicated. I would not set CXXFLAGS_PROGRAMS inside each if block, just set the `HAS_CPP17` flag, and then at the end (pseudocode)

```
if HAS_CPP17 
  CXXFLAGS_PROGRAM ?= -std=c++17
else
  CXXFLAGS_PROGRAM ?= -std=c++1y
  if NOT STAN_USE_CPP14
    warning()